### ISSUE TITLE
Remove or relax the length limitation of Uri

### ISSUE THREAD

**Issue by hez2010** [dotnet/runtime#96544]
The `data` is a valid scheme which can carry some data other than path in a `Uri`. For example, we can save an image as base64 data in a `Uri`:

```csharp
var uri = new Uri("data:image/jpeg;base64,dGVzdA==");
```

However, `Uri` in .NET has a maximum length limitation of `uint.MaxValue` which makes it unable to save some large data (especially, length of base64 more than 65496).

It's common that a HTML contains an `image` tag which use base64 encoded data for its `src` attribute:

```html
<image src="data:image/jpeg;base64,dGVzdA==" />
```

However, we are unable to get the `src` as a `Uri` if the length of `src` is too large, for example, an 1 mb image.

```csharp
new Uri("data:image/jpeg;base64," + Enumerable.Repeat("a", 70000).Aggregate((s, n) => $"{s}{n}"))
```

The above code will throw a `UriFormatException` with `Invalid URI: The Uri string is too long.`.

I would like to suggest remove this limitation, or relax it, for example, relax the limitation from `uint.MaxValue` to `int.MaxValue`.

**Comment by colejohnson66** [comment]
https://github.com/dotnet/runtime/blob/22068a8f96d6d1c01b26db70fd24a433e1fdd5ef/src/libraries/System.Private.Uri/src/System/Uri.cs#L1911-L1912

What's the purpose of there even being a cap on the length? [RFC 3986](https://www.rfc-editor.org/rfc/rfc3986) doesn't mention anything about a limit needing to exist, and there's no comment as to why.

**Comment by MihaZupan** [comment]
The same discussion from a few years ago: [Uri rejects otherwise valid strings with length >= 65520 (#1857)](https://github.com/dotnet/runtime/issues/1857)

Essentially:
- `Uri` is practically useless for `data:` uris, as it provides no meaningful validation, and exposes no utility methods for extracting the data from the format
- Having a length limit makes sense in most cases (a 64 KB url is huge for HTTP), but obviously unfortunate for `data:`
- This isn't a random constant. The limitation (mainly?) came from the fact that Uri internally stores offsets as `ushort`s (hence the ~2^16 length limit). Lifting it would mean [more work](https://github.com/dotnet/runtime/issues/1857#issuecomment-592651720) to make sure other parts of Uri play nice

Ideally we would have a dedicated type for dealing with data Uris - https://github.com/dotnet/runtime/issues/85164#issuecomment-1557163136.

But we don't have a good answer for cases where you're stuck with using `Uri` by existing APIs, like in your case, so I'm hesitant to close this as wont-fix as we did with #1857.

Moving to Future for now to let others comment/upvote if they run into the same problem

**Comment by dersia** [comment]
@MihaZupan this would be really important for the Azure Ai sdk and semantic kernel and I would love for this to move forward. can we do something to move this forward?

I thought about your suggestion to use a new type like proposed in https://github.com/dotnet/runtime/issues/85164 but the problem that I see with that is for apis/sdks like Azure SDK that are built from open-api-specs using tools like autorest. the type in the spec just says `url` and what would you map it to? `System.Uri` or `System.DataUri`.

so from that point of view I think it makes much mir sense to move this proposal forward and make `System.Uri` to work with DataUri's. and in fact it does work with DataUri's already, it is just that because of the size limitations it won't work with bigger DataUri's.

also adding a new type would bring up a weird state where you can use `System.Uri` as long as you stay within bound and have to switch to `System.DataUri` when you exceed this limit. so there would be no clear recommendation which one to use + this isn't breaking.

**Comment by mikepizzo** [comment]
Note: this is not just for data: uris.

We have scenarios in which users create really long queries (for example, return instances that match this set of ids, where the number of ids can be in the thousands).

In OData, we have added a pattern for adding /$query to the resource path and using POST to pass the query string in the body of the request to work around limitations with HTTP stacks.  However, internally we still build full URIs for the request (for example, in order to generate a "nextLink") and run into this limitation.

See, for example, [OData#1293](https://github.com/OData/AspNetCoreOData/issues/1293)

**Comment by hez2010** [comment]
Yeah. This issue has to be fixed as internally we are creating an instance of `Uri` even if you are using the `string` overload in `HttpMessageRequest`.
You have no way to workaround this issue if you use `HttpClient`.
